atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java

```diff
@@ -652,6 +634,94 @@ public String toString() {
+                    .filter(cells -> !cells.isEmpty())
+                    .flatMap(Collection::stream)
+                    .collectToMultimap(LinkedListMultimap::create);
+
+            result.putAll(partialResult);
+
+            query = KeyedStream.stream(Multimaps.asMap(partialResult))
+                    .map((row, cells) -> keyPredicate(row, getNextLexicographicalSlicePredicate(cells)))
+                    .values()
+                    .collect(Collectors.toList());
+        }
+
+        ValueExtractor extractor = new ValueExtractor(metricsManager, Maps.newHashMapWithExpectedSize(result.size()));
+        extractor.extractResults(Multimaps.asMap(result), startTs, ColumnSelection.all());
+        return extractor.asMap();
+    }
+
+    private static KeyPredicate keyPredicate(ByteBuffer row, SlicePredicate predicate) {
+        return new KeyPredicate().setKey(row).setPredicate(predicate);
+    }
+
+    private static SlicePredicate allPredicateWithLimit(int limit) {
+        return SlicePredicates.create(Range.ALL, Limit.of(limit));
+    }
+
+    private Map<ByteBuffer, List<ColumnOrSuperColumn>> fetchCellsForKeyPredicates(final InetSocketAddress host,
+            final TableReference tableRef,
+            List<KeyPredicate> query,
+            final long startTs) throws Exception {
+        return clientPool.runWithRetryOnHost(
+                host,
+                new FunctionCheckedException<CassandraClient, Map<ByteBuffer, List<ColumnOrSuperColumn>>, Exception>() {
+                    @Override
+                    public Map<ByteBuffer, List<ColumnOrSuperColumn>> apply(CassandraClient client) throws Exception {
+
+                        if (log.isTraceEnabled()) {
+                            log.trace("Requesting {} cells from {} starting at timestamp {} on {} "
+                                            + "as part of fetching cells for key predicates.",
+                                    SafeArg.of("cells", query.size()),
+                                    LoggingArgs.tableRef(tableRef),
+                                    SafeArg.of("startTs", startTs),
+                                    SafeArg.of("host", CassandraLogHelper.host(host)));
+                        }
+
+                        Map<ByteBuffer, List<List<ColumnOrSuperColumn>>> results = wrappingQueryRunner.multiget_multislice(
+                                "getRows", client, tableRef, query, readConsistency);
+
+                        return Maps.transformValues(results,
+                                lists -> Lists.newArrayList(Iterables.concat(lists)));
+                    }
+
+                    @Override
+                    public String toString() {
+                        return "multiget_multislice(" + host + ", " + tableRef +
+                                + query.size() + " cells" + ")";
```

Comment:
nit: `... + " cells)";` rather than two separate strings

Reply:
And we probably want a comma before saying the number of cells. The idea is to get something like `multiget_multislice(host, table, 1000 cells)`